src/controllers.lisp

```diff
@@ -116,15 +123,41 @@
     (let ((*draw-mode* nil))
       (call-next-method))))
 
-(defmethod kit.sdl2:textinput-event :after ((instance sketch) timestamp text)
-  (on-text instance text))
+(defmethod kit.sdl2:textinput-event :after ((instance sketch-window) timestamp text)
+  (on-text (sketch instance) text))
 
-(defmethod kit.sdl2:keyboard-event :after ((instance sketch) state timestamp repeat-p keysym)
+(defmethod kit.sdl2:keyboard-event :after ((instance sketch-window) state timestamp repeat-p keysym)
   (when (not repeat-p)
-    (on-key instance
+    (on-key (sketch instance)
             ;; Removing the ugly "SCANCODE-" prefix from the keyword
             ;; symbol that denotes the button.
             (intern (subseq (symbol-name (sdl2:scancode keysym))
                             +scancode-prefix-length+)
                     (find-package "KEYWORD"))
             state)))
+
+;;; Backward compatibility.
+;; Previously, the main `sketch` class inherited from
+;; `kit.sdl2:gl-window`, and input was handled by specialising on methods from
+;; sdl2kit. So we need to forward sdl2kit input calls to the `sketch` class for
+;; old sketches that rely on that approach.
+(defmacro define-sdl2-forward (name (&rest args) &optional already-defined?)
+  `(progn
+     ;; An empty method so we don't get an error if we try to forward
+     ;; when the user hasn't defined it.
+     (defmethod ,name ((w sketch) ,@args))
+     ,@(when (not already-defined?)
+         `((defmethod ,name ((w sketch-window) ,@args)
+             (,name (sketch w) ,@args)
+             (call-next-method))))))
+(define-sdl2-forward kit.sdl2:mousebutton-event (state timestamp button x y) t)
+(define-sdl2-forward kit.sdl2:mousemotion-event (timestamp button-mask x y xrel yrel) t)
+(define-sdl2-forward kit.sdl2:close-window () t)
```

Comment:
`already-defined?` argument here should not be `t`: the "direct" method is not defined, only a `:before` and `:after` methods.

Reply:
Actually, I think that `close-window` should be defined separately similarly to `render` -- user might be calling it directly on sketch's instance to close it.

Reply:
Hm, this one is tricky: on the one hand, user might want to close the sketch manually with calling `(kit.sdl2:close-window sketch)`; on the other hand, user might want to define a `:before` / `:after` / `:around` method, to add cleanup.

If `kit.sdl2:close-window` on `((sketch sketch))` calls `(kit.sdl2:close-window (sketch-%window sketch))`, then when a window is closed by `kit.sdl2` (for example when window is closed by OS), user-defined cleanup won't be executed.

If `kit.sdl2:close-window` on `((sketch sketch-window))` calls `(kit.sdl2:close-window (sketch window))`, the window still can't be closed by a direct call `(kit.sdl2:close-window sketch)` because the actual closing done by `kit.sdl2` won't be executed.